### PR TITLE
Add android QR detection sample

### PR DESCRIPTION
**Pull request by AleksandrPanov**
Android QR detection sample was added.

image
image
image

current interface:
image

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionActivity.java
Comment (dkurt) [inline samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionActivity.java:1]: Why not named as  `QRdetectionActivity.java`?

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+     * the tiles as specified by mIndexes array
+     */
+    public synchronized Mat qrFrame(Mat inputPicture, boolean tryDecode) {
+        List<String> decoded_info = new ArrayList<String>();
+        Mat points = new Mat();
+        boolean result;
+        if (tryDecode)
+            result = detector.detectAndDecodeMulti(inputPicture, decoded_info, points);
+        else
+            result = detector.detectMulti(inputPicture, points);
```
Comment (dkurt) [inline samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java:43]: I suppose `detectMulti` has worse performance comparing to `detect`. So shall we add an item to switch between single/multiple QR codes detection?

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        else
+            detector = new QRCodeDetector();
+    }
+
+    /* this method to be called from the outside. it processes the frame and shuffles
+     * the tiles as specified by mIndexes array
+     */
+    public synchronized Mat qrFrame(Mat inputPicture, boolean tryDecode) {
+        List<String> decoded_info = new ArrayList<String>();
+        Mat points = new Mat();
```
Comment (dkurt) [inline samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java:38]: If I'm not mistaken, it's better to call `.release` method for every OpenCV's object explicitly.

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            result = detector.detectMulti(inputPicture, points);
+        if (result) {
+            for (int i = 0; i < points.rows(); i++) {
+                Point start = new Point(0., 0.);
+                for (int j = 0; j < points.cols(); j++) {
+                    Point pt1 = new Point(points.get(i, j));
+                    Point pt2 = new Point(points.get(i, (j + 1) % 4));
+                    start.x += pt1.x / 4.;
+                    start.y += pt1.y / 4.;
+                    Imgproc.line(inputPicture, pt1, pt2, new Scalar(255, 0, 0), 3);
```
Comment (dkurt) [inline samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java:52]: [polylines()](https://docs.opencv.org/5.x/d6/d6e/group__imgproc__draw.html#gaa3c25f9fb764b6bef791bf034f6e26f5)

File: samples/android/qr-detection/res/drawable/icon.png
Comment (dkurt) [inline samples/android/qr-detection/res/drawable/icon.png:1]: Can gradle reuse an image from parent projects or download by URL?

File: samples/android/qr-detection/res/values/strings.xml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="app_name">OCV QR Detection</string>
```
Comment (dkurt) [inline samples/android/qr-detection/res/values/strings.xml:3]: I suggest `OpenCV QR Detection`

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    public synchronized Mat qrFrame(Mat inputPicture, boolean tryDecode) {
+        List<String> decoded_info = new ArrayList<String>();
+        Mat points = new Mat();
+        boolean result;
+        if (tryDecode)
+            result = detector.detectAndDecodeMulti(inputPicture, decoded_info, points);
+        else
+            result = detector.detectMulti(inputPicture, points);
+        if (result) {
+            for (int i = 0; i < points.rows(); i++) {
```
Comment (asmorkalov) [inline samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java:45]: I propose to extract rendering as method and introduce Scalars for colors as named constatnts to not re-create them each time.

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class is a controller for puzzle game.
+ * It converts the image from Camera into the shuffled image
+ */
+public class QrdetectionProcessor {
```
Comment (asmorkalov) [inline samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java:23]: QR with capital letters look better.

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java (hunk 3 above)
Comment (AleksandrPanov) [reply samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java:52]: We need use too complex code with type cast in this case:
1st sample:

```
                MatOfPoint2f tmpPoints = new MatOfPoint2f(points.row(i));
                MatOfPoint tmpIntPoints = new MatOfPoint();
                tmpPoints.convertTo(tmpIntPoints, CvType.CV_32S);
                List<MatOfPoint> tmpListPoints = Arrays.asList(tmpIntPoints);
                Imgproc.polylines(img, tmpListPoints, true, new Scalar(255, 255, 255), 3);
```

2nd sample:

```
                Mat m1 = new Mat();
                points.row(i).convertTo(m1, CvType.CV_32S);
                MatOfPoint m2 = new MatOfPoint(m1);
                List<MatOfPoint> m3 = new ArrayList<>();
                m3.add(m2);
                Imgproc.polylines(img, m3, true, new Scalar(255, 255, 255), 3);
```

I want to use just `lines`.

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionActivity.java
Comment (AleksandrPanov) [reply samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionActivity.java:1]: was renamed

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java (hunk 1 above)
Comment (AleksandrPanov) [reply samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java]: added `Use multi detect/decode` flag

File: samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java (hunk 2 above)
Comment (AleksandrPanov) [reply samples/android/qr-detection/src/org/opencv/samples/qrdetection/QrdetectionProcessor.java]: add .`release()` for `points`

File: samples/android/qr-detection/res/drawable/icon.png
Comment (AleksandrPanov) [reply samples/android/qr-detection/res/drawable/icon.png:1]: I do not know how to do this

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**